### PR TITLE
Problem: global _sdk pointer is unsafe which likely makes game crashe…

### LINKED ISSUE #307: Problem: global _sdk pointer is unsafe which likely makes game crashes

### PR DESCRIPTION
**Pull request by damoncro**
…s (fix #307)

Solution:
- Remove global pointer
- Pass the PlayCppSdk pointer to UserWalletConnectCallback during initialization
- Replace sendEvent with lambda function

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-com/play-unreal-plugin/blob/main/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest main? 
- [ ] Have you checked your code compiles in Unreal Engine 4 and 5?
- [ ] Have you checked your basic code formatting and style is fine? ([using this Linter plugin](https://ue4-style-guide.readthedocs.io/en/latest/gettingstarted.html))
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the plugin version number and documented your changes in the [CHANGELOG.md](https://github.com/crypto-com/play-unreal-plugin/blob/main/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-com/play-unreal-plugin/blob/main/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-com/play-unreal-plugin/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: Source/CronosPlayUnreal/Private/PlayCppSdkActor.cpp [hunk 1]
```diff
@@ -23,36 +23,128 @@ using namespace com::crypto::game_sdk;
 
 ::com::crypto::game_sdk::WalletconnectClient *APlayCppSdkActor::_coreClient =
     NULL;
-APlayCppSdkActor *APlayCppSdkActor::_sdk = NULL;
 
 class UserWalletConnectCallback : public WalletConnectCallback {
+  private:
+    APlayCppSdkActor *PlayCppSdkPtr;
```
Comment (leejw51crypto) [inline Source/CronosPlayUnreal/Private/PlayCppSdkActor.cpp:29]: acessing PlayCppSdkPtr from raw thread is unsafe.
```
#include "Async/Async.h"

UObject* MyUnrealObject; // UObject pointer you want to access

// Your raw thread function
void MyRawThreadFunction()
{
    // ... your code here ...

    // Use AsyncTask to execute code on the game thread
    AsyncTask(ENamedThreads::GameThread, [&]()
    {
        if (MyUnrealObject != nullptr)
        {
            // Access the UObject safely on the game thread
            // ... your code here ...
        }
    });

    // ... more code ...
}

```


and store raw pointer as weakpointer, and check `isvalid` before hand
```
TWeakObjectPtr<UClass> MyWeakClass;

// Check if the UObject is still valid before using it
if (MyWeakClass.IsValid())
{
    // Use the UObject safely
}
```

Comment (leejw51crypto) [reply Source/CronosPlayUnreal/Private/PlayCppSdkActor.cpp:29]: so only game thread can access any pointer safely.

File: Source/CronosPlayUnreal/Private/PlayCppSdkActor.cpp [hunk 2]
```diff
@@ -23,36 +23,128 @@ using namespace com::crypto::game_sdk;
...
+    const WalletConnectSessionInfo &sessioninfo) const {
+    UE_LOG(LogTemp, Log, TEXT("user c++ onConnected"));
+    if (PlayCppSdkPtr) {
+        APlayCppSdkActor *PlayCppSdk =
+            PlayCppSdkPtr; // we need it to create a new varialbe for lambda
+        // We could not Call PlayCppSdkPtr's function direclty, bcause it is not
+        // in game thread. Instead, we can create AsyncTask and pass the
+        // PlayCppSdkPtr and sessioninfo into it, then call the function in game
+        // thread.
+        AsyncTask(ENamedThreads::GameThread, [PlayCppSdk, sessioninfo]() {
```
Comment (leejw51crypto) [inline Source/CronosPlayUnreal/Private/PlayCppSdkActor.cpp:53]: confirmed.
calling delegate from asynctask, which making caller side in unreal game thread

File: Source/CronosPlayUnreal/Private/PlayCppSdkActor.cpp (hunk 1 above)
Comment (damoncro) [reply Source/CronosPlayUnreal/Private/PlayCppSdkActor.cpp:29]: Yes, it is not 100% fix the pointer sharing issue. I've marked it TODO:
https://github.com/cronos-labs/play-unreal-plugin/pull/308/files#diff-097b9056119eecdfb8e1f150f586de6b3ed6ef617853a734a6f0d94006274fa1R516-R517

It is unsafe but worked at this moment. Much better than the original _sdk global pointer.

**Review by leejw51crypto — APPROVED** [review]
lgtm
we can do fine-tuning as another pr

**Pull request merged**